### PR TITLE
Change the for-loop desugar so the `break` does not affect type inference. Fixes #42618

### LINKED ISSUE #42618: Rust 1.19 regression, minions 0.2.8, type mismatch
https://github.com/dbeck/acto-rs

```
commit 5311c872dd88d3bf034e023f501971c711dee41c
Author: David Beck <david.beck.priv@gmail.com>
Date:   Mon Aug 29 23:04:00 2016 +0100

    sanitize Observer interface
```

```
rustc 1.19.0-beta.1 (a87984118 2017-06-06)
binary: rustc
commit-hash: a879841184d6fe89b819425fde1568b50a54543f
commit-date: 2017-06-06
host: x86_64-unknown-linux-gnu
release: 1.19.0-beta.1
LLVM version: 4.0

```

```
brian@ip-10-145-43-250:~/dev/acto-rs⟫ cargo +beta test
   Compiling minions v0.2.7 (file:///mnt2/dev/acto-rs)
error[E0308]: mismatched types
   --> src/main.rs:147:14
    |
147 |       sum += i;
    |              ^ expected i32, found ()
    |
    = note: expected type `i32`
               found type `()`

error: aborting due to previous error(s)

error: Could not compile `minions`.

To learn more, run the command again with --verbose.
```

cc @dbeck

### PR DESCRIPTION
**Pull request by Zoxc**
Rewrite the `for` loop desugaring to avoid contaminating the inference results. Under the older desugaring, `for x in vec![] { .. }` would erroneously type-check, even though the type of `vec![]` is unconstrained. (written by @nikomatsakis)

### CODE REVIEW

**Comment by arielb1** [comment]
This looks like a rather fragile area of code. I'm ok with the changes, but I'll rather r? @nikomatsakis

**Comment by nikomatsakis** [comment]
@Zoxc can we add a run-pass test checking the order of drops with comments explaining the **expected** order?

**Comment by nikomatsakis** [comment]
We probably would need this to be in relnotes, presuming it lands. Tagging it as such.

**Comment by nikomatsakis** [comment]
@Zoxc can you also add a `run-pass` test case like this? (It fails for me on nightly/beta, not on stable)

```rust
fn main() {
    let mut sum = 0;
    for i in Vec::new() {
        sum += i;
    }
}
```

Similarly a `compile-fail` test like this, which also fails on stable, but works on beta/nightly:

```rust
fn main() {
    for i in Vec::new() { } //~ ERROR type annotations needed
}
```

r=me with those tests

**Comment by nikomatsakis** [comment]
Removing relnotes, I didn't realize that this behavior had never contaminated stable.

**Comment by brson** [comment]
https://github.com/rust-lang/rust/pull/42265 needs to be reverted as this is applied

**Comment by nikomatsakis** [comment]
I took the liberty of adding a few comments to the test for posterity.

**Comment by nikomatsakis** [comment]
@bors r-

oops, forgot one test.

**Comment by nikomatsakis** [comment]
tidy error:

```
[00:02:09] tidy error: /checkout/src/librustc/hir/lowering.rs:2192: line longer than 100 chars
```

File: src/librustc/hir/lowering.rs [hunk 1]
```diff
@@ -2170,11 +2170,13 @@ impl<'a> LoweringContext<'a> {
                 //     let result = match ::std::iter::IntoIterator::into_iter(<head>) {
                 //       mut iter => {
                 //         [opt_ident]: loop {
-                //           let <pat> = match ::std::iter::Iterator::next(&mut iter) {
-                //             ::std::option::Option::Some(val) => val,
+                //           let mut _next;
```
Comment (nikomatsakis) [inline src/librustc/hir/lowering.rs]: Does the `_next` name disable a "unused mut" warning?

Comment (Zoxc) [inline src/librustc/hir/lowering.rs]: It does

**Pull request merged**